05-data-structures-part2.Rmd

```diff
@@ -1,245 +1,132 @@
-> you expect?
->
-
-Each column in the data frame is simply a list element, which is why
-when you ask for the `length` of the data frame, it tells you the
-number of columns. If you actually want the number of rows, you can
-use the `nrow` function.
-
-We can add columns or rows to a data.frame using `cbind` or `rbind`
-(these are the two-dimensional equivalents of the `c` function):
-
-## Adding columns to a data frame
-
-To add a column we can use `cbind`:
+We can then add this as a column via:
 
 ```{r}
-df <- cbind(df, 6:1)
-df
+cats <- cbind(cats,  newCol)
 ```
 
-Note that R automatically names the column. We may decide to change
-the name by assigning a value using the `names` function:
+Why didn't this work? Of course, R wants to see one element in our new column for every row in the table:
 
 ```{r}
-names(df)[4] <- 'SixToOne'
+cats
+newCol <- c(4,5,8)
+cats <- cbind(cats, newCol)
+cats
 ```
 
-We can also provide a name when we add the column:
+Our new column has appeared, but it's got that ugly name at the top; let's give it something a little easier to understand:
 
 ```{r}
-df <- cbind(df, caps=LETTERS[1:6])
-df
+names(cats)[4] <- 'age'
 ```
 
-(`LETTERS` and `letters` are built-in constants.)
-
-
-## Adding rows to a data frame 
-
-To add a row we use `rbind`:
+Now how about adding rows - in this case, we saw last time that the rows of a data.frame are made of lists:
 
 ```{r}
-df <- rbind(df, list("g", 11, 42, 0, "G"))
+newRow <- list("tortoiseshell", 3.3, TRUE, 9)
+cats <- rbind(cats, newRow)
 ```
 
-Note that we add the row as a list, because we have multiple types
-across the columns. Nevertheless, this doesn't work as expected! What
-do the error messages tell us?
-
-It appears that R was trying to append "g" and "G" as factor
-levels. Why? Let's examine the first column.  We can access a column
-in a `data.frame` by using the `$` operator.
-
+Another thing to look out for has emerged - when R creates a factor, it only allows whatever is originally there when our data was first loaded, which was 'black', 'calico' and 'tabby' in our case. Anything new that doesn't fit into one of its categories is rejected as nonsense, until we explicitly add that as a *level* in the factor:
 
 ```{r}
-class(df$id)
-str(df)
+levels(cats$coat)
+levels(cats$coat) <- c(levels(cats$coat), 'tortoiseshell')
+cats <- rbind(cats, list("tortoiseshell", 3.3, TRUE, 9))
 ```
 
-When we created the data frame, R automatically made the first
-and last columns into factors, not character vectors. There are no
-pre-existing levels "g" and "G", so the attempt to add these values
-fails. A row was added to the data frame, only there are missing
-values in the factor columns:
+Alternatively, we can change a factor column to a character vector; we lose the handy categories of the factor, but can subsequently add any word we want to the column without babysitting the factor levels:
 
 ```{r}
-df
+str(cats)
+cats$coat <- levels(cats$coat)[cats$coat]
```

Comment:
This is elegant in the extreme, and nicely illustrates why the secret numeric nature of factors is a feature, not a bug. 
However wouldn't it be clearer to use `as.character`, since we've already introduced the gang of coercion functions?